content/lessons/advanced-topics/dynamodb-streams.md

```diff
@@ -7,4 +7,150 @@ date: "01/01/2018"
+
+The probably most common way of handling DynamoDB stream items is by having a Lambda function listen to the stream. You can do that by using the Lambda console or by using a `AWS::Lambda::EventSourceMapping` resource in your Cloudformation template. As part of that action you need to decide on the `BatchSize` the underlying infrastructure will pass into your function. Your function need to be able to handle that. If you configure a batch size of 150 you will receive an array of up to 150 items from the DynamoDB stream. As usual in Lambda, you need to think about your business logic, also in terms of memory and timeout settings to come up with a batch size that works best.
+
+A DynamoDB stream item would look similar to this (with a couple of attributes removed for clarity):
+
+```
+{
+   "Records":[
+      {
+         "eventID":"1",
+         "eventName":"INSERT",
+         "dynamodb":{
+            "Keys":{
+               "MyPartitionKey":{
+                  "S":"f616ccf8-94da-48b5-9c8e-8ea8bc701c16"
+               }
+            },
+            "NewImage":{
+               "MyData":{
+                  "S":"This is a new item!"
+               },
+               "MyPartitionKey":{
+                  "S":"f616ccf8-94da-48b5-9c8e-8ea8bc701c16"
+               }
+            }
+         }
+      },
+      {
+         "eventID":"2",
+         "eventName":"MODIFY",
+         "dynamodb":{
+            "Keys":{
+               "MyPartitionKey":{
+                  "S":"899ff269-47b4-4aca-aac1-a7da2c528512"
+               }
+            },
+            "NewImage":{
+               "MyData":{
+                  "S":"This is the new attribute data"
+               },
+               "MyPartitionKey":{
+                  "S":"899ff269-47b4-4aca-aac1-a7da2c528512"
+               }
+            },
+            "OldImage":{
+               "Message":{
+                  "S":"This is the old attribute data"
+               },
+               "MyPartitionKey":{
+                  "S":"899ff269-47b4-4aca-aac1-a7da2c528512"
+               }
+            }
+         }
+      },
+      {
+         "eventID":"3",
+         "eventName":"REMOVE",
+         "dynamodb":{
+            "Keys":{
+               "MyPartitionKey":{
+                  "S":"14cb3e02-d5f4-4c4a-b7d5-2811c7055552"
+               }
+            },
+            "OldImage":{
+               "MyData":{
+                  "S":"Some data"
+               },
+               "MyPartitionKey":{
+                  "S":"14cb3e02-d5f4-4c4a-b7d5-2811c7055552"
+               }
+            }
+         }
+      }
+   ]
+}
+```
+
+This is an array of three items from a stream. The lambda code now can react to whatever attribute is relevant.
+
+**eventID** is the unique ID of the stream item. Use this for logging, tracing, de-duplication (think idempotency!)
```

Comment:
Could you add some more details about the de-duplication bit? For example by moving the **Idempotency** explanation over here.